src/ApplicationSampler.cpp

```diff
@@ -0,0 +1,251 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include <map>
+
+#include "ApplicationSampler.hpp"
+#include "ProfileSampler.hpp"
+#include "ProfileIOSample.hpp"
+#include "EpochRuntimeRegulator.hpp"
+#include "Exception.hpp"
+
+namespace geopm
+{
+    class ApplicationSamplerImp : public ApplicationSampler
+    {
+        public:
+            ApplicationSamplerImp();
+            virtual ~ApplicationSamplerImp() = default;
+            void time_zero(const geopm_time_s &start_time) override;
+            void update_records(void) override;
+            std::vector<m_record_s> get_records(void) const override;
+            std::vector<std::string> get_names(uint64_t name_key) const override;
+            std::vector<uint64_t> per_cpu_hint(void) const override;
+            std::vector<double> per_cpu_progress(void) const override;
+
+            void set_sampler(std::shared_ptr<ProfileSampler> sampler) override;
+            std::shared_ptr<ProfileSampler> get_sampler(void) override;
+            void set_regulator(std::shared_ptr<EpochRuntimeRegulator> regulator) override;
+            std::shared_ptr<EpochRuntimeRegulator> get_regulator(void) override;
+            void set_io_sample(std::shared_ptr<ProfileIOSample> io_sample) override;
+            std::shared_ptr<ProfileIOSample> get_io_sample(void) override;
+        private:
+            struct m_process_s {
+                uint64_t epoch_count;
+                uint64_t hint;
+            };
+            std::map<int, m_process_s>::iterator get_process(int process);
+            std::shared_ptr<ProfileSampler> m_sampler;
+            std::shared_ptr<EpochRuntimeRegulator> m_regulator;
+            std::shared_ptr<ProfileIOSample> m_io_sample;
+            struct geopm_time_s m_time_zero;
+            std::vector<geopm_prof_message_s> m_sample_cache;
+            std::vector<m_record_s> m_record_buffer;
+            std::map<int, m_process_s> m_process_map;
+    };
+
+
+    ApplicationSampler &ApplicationSampler::application_sampler(void)
+    {
+        static ApplicationSamplerImp instance;
+        return instance;
+    }
+
+
+    ApplicationSamplerImp::ApplicationSamplerImp()
+    {
+
+    }
+
+
+    void ApplicationSamplerImp::time_zero(const geopm_time_s &start_time)
+    {
+        m_time_zero = start_time;
+    }
+
+
+    void ApplicationSamplerImp::update_records(void)
```

Comment:
Function should probably be broken down into smaller private methods for each different kind of geopm_message_s.